CRM/Mailing/Page/Url.php

```diff
@@ -0,0 +1,104 @@
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+
+/**
+ * Redirects a user to the full url from a mailing url.
+ *
+ * General Usage: civicrm/mailing/url?qid={event_queue_id}&u={url_id}
+ *
+ * Additional arguments may be handled by extractPassthroughParameters().
+ */
+class CRM_Mailing_Page_Url extends CRM_Core_Page {
+
+  /**
+   * Redirect the user to the specified url.
+   *
+   * @throws \CRM_Core_Exception
+   */
+  public function run() {
+    $queue_id = CRM_Utils_Request::retrieveValue('qid', 'Integer');
+    $url_id = CRM_Utils_Request::retrieveValue('u', 'Integer', NULL, TRUE);
+    $url = CRM_Mailing_Event_BAO_TrackableURLOpen::track($queue_id, $url_id);
+    $query_string = $this->extractPassthroughParameters();
+
+    if (strlen($query_string) > 0) {
+      // Parse the url to preserve the fragment.
+      $pieces = parse_url($url);
+
+      if (isset($pieces['fragment'])) {
+        $url = str_replace('#' . $pieces['fragment'], '', $url);
+      }
+
+      // Handle additional query string params.
+      if ($query_string) {
+        if (stristr($url, '?')) {
+          $url .= '&' . $query_string;
+        }
+        else {
+          $url .= '?' . $query_string;
+        }
+      }
+
+      // slap the fragment onto the end per URL spec
+      if (isset($pieces['fragment'])) {
+        $url .= '#' . $pieces['fragment'];
+      }
+    }
+    CRM_Utils_System::redirect($url, [
+      'for' => 'civicrm/mailing/url',
+      'queue_id' => $queue_id,
+      'url_id' => $url_id,
+    ]);
+  }
+
+  /**
+   * Determine if this request has any valid pass-through parameters.
+   *
+   * Under CRM-7103 (v3.3), all unrecognized query-parameters (besides qid/u) are passed
+   * through as part of the redirect. This mechanism is relevant to certain
+   * customizations (eg using `hook_alterMailParams` to append extra URL args)
+   * but does not matter for normal URLs.
+   *
+   * The functionality seems vaguely problematic (IMHO) - especially now that
+   * 'extern/url.php' is moving into the CMS/Civi router ('civicrm/mailing/url').
+   * But it's the current protocol.
+   *
+   * A better design might be to support `hook_alterRedirect` in the CiviMail
+   * click-through tracking. Then you don't have to take any untrusted inputs
+   * and you can fix URL mistakes in realtime.
+   *
+   * @return string
+   * @link https://issues.civicrm.org/jira/browse/CRM-7103
+   */
+  protected function extractPassthroughParameters():string {
+    $config = CRM_Core_Config::singleton();
+
+    $query_param = $_GET;
+    unset($query_param['qid']);
+    unset($query_param['u']);
+    unset($query_param[$config->userFrameworkURLVar]);
+    if ($config->userFramework === 'WordPress') {
+      // Ugh
+      unset($query_param['page']);
+      unset($query_param['noheader']);
+    }
```

Comment:
When I first tested on a `wpmaster` site, there were still more noisy GET params being passed through (e.g. a link to `https://www.google.com` winds up as `https://www.google.com?noheader=1`). This extra conditional resolved that.

But it's also ugly. TBH I'd be happier to rip out `extractPassthroughParameters()` and make a small BC-break:

* The notion in https://issues.civicrm.org/jira/browse/CRM-7103 / `extractPassthroughParameters()` is not well-defined across different routing systems.
* This PR makes a better API available for the same use-case (`hook_civicrm_alterRedirect`).
* I skimmed `hook_civicrm_alterMailParams` consumers in `universe` and couldn't spot any doing URL manipulation on CiviMail content. (The fuzion `transactional` ext was close, but it skips CiviMail.)
* If one is actually relying on `hook_civicrm_alterMailParams` for URL manipulation, then you can buy some time for a transition -- set `defaultExternUrl=standalone` and you get the old behavior.

Reply:
@totten I'm pretty sure that [this issue on Lab](https://lab.civicrm.org/dev/wordpress/-/issues/49) will have an impact here - because bleeding edge WordPress has changed the way in which the `page` query var is used and it now assumes that this is the the desired path and redirects accordingly. We're going to have to audit CiviCRM's usage of this query var in a WordPress context and strip out any usage of it on the front-end. This should lead to `unset($query_param['page']);` being unnecessary.

Reply:
OK, well, `?page=` won't be missed. :)

Since coordinating the timelines on these changes may be tricky, I guess the implication here is... if #17312 does keep `extractPassthroughParameters()`, then eventually we can remove the `unset($query_param['page')`. For the moment, from a WP-integration perspective, there's no harm in keeping the `unset()`.